ls1intum/Artemis
# [Enhancement] Replace fork repositories with copy

### Checklist

### Motivation and Context


We want to get rid of forking repositories at all and instead copy them. In an earlier PR we had changed the behavior of forking during an import of an exercise to use copying. This is why we want to have it consistent throughout Artemis. 
 
### Description

When a student starts a new exercise the source repository doesn't get forked anymore but cloned. 
The process is: 
The source repository gets cloned into the students unique folder of the course exercise. Now the remote url of the student's repo gets changed to the correct target remote and we push the repo. By doing this, the student's repo on the VC system gets updated with the template/source code.  
The mechanism to import programming exercises is changed to the same behaviour.

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise
4. Open the Bitbucket/GitLab page of the exercise, to verify the next steps
5. Start the exercise
5.1 Check in the VC page that the newly created repo is cloned and not forked
6. Participate in the exercise and submit some code
6.1 Check that the template repository was not affected and was not changed
**New:** 6.2. Clone the repo and commit+push changes from your local computer (to verify the permissions are correct)
7. Import the exercise and check that the repos (template, solution, tests) are cloned and not forked
7.1 Open the template repo of the imported exercise and submit some code.
7.2 Verify that the original template repo was not changed
8. We also need to test if the participation of an programming exercise in an exam is still possible

Note: we need to test on both server environments (ts1/3 + ts2)

### Test Coverage

The tests were refactored to take the cloning of repos into account. But the test coverage stays roughly the same.
- ParticipationService.java: 78%
- ProgrammingExerciseImportService: 96%
- AbstractVersionControlService: 92%
- GitService: 74%
- GitLabService: 70%
- BitbucketService: 77%

### Screenshots



Different code after submitting of code:
Template repo:
![image](https://user-images.githubusercontent.com/33484318/105217344-04bfef00-5b54-11eb-9b8b-a28165456a86.png)
Submitted different code in started exercise:
![image](https://user-images.githubusercontent.com/33484318/105217621-64b69580-5b54-11eb-86c0-470e18063d72.png)


Verify it is a cloned and not a forked symbol:
![image](https://user-images.githubusercontent.com/33484318/105217158-cc201580-5b53-11eb-9259-79546aa8be1d.png)

Pushing to the repo after the (exam) due date should be forbidden:
![image](https://user-images.githubusercontent.com/33484318/105704427-136d2400-5f0f-11eb-88fb-9ca1f8358c75.png)
